### PR TITLE
Rewrite repo updater Store

### LINKED ISSUE #14714: repo updater store: Simplify interface
The repo updater specific store should no longer be an interface but a concrete type, specifically:

```go
// Concrete type with baked-in observability
type Store struct{}

ListExternalRepoSpecs(context.Context) (map[api.ExternalRepoSpec]struct{}, error)
UpsertRepos(ctx context.Context, repos ...*types.Repo) error
UpsertSources(ctx context.Context, inserts, updates, deletes map[api.RepoID][]SourceInfo) error
SetClonedRepos(ctx context.Context, repoNames ...string) error
CountNotClonedRepos(ctx context.Context) (uint64, error)
CountUserAddedRepos(ctx context.Context) (uint64, error)
EnqueueSyncJobs(ctx context.Context, ignoreSiteAdmin bool) error
```

The store must be moved to `cmd/repo-updater/internal/store.go`

Extracted from: https://github.com/sourcegraph/sourcegraph/issues/14092

### PR DESCRIPTION
**Pull request by asdine**
This PR adds a new store to repo updater. It's a concrete type provides methods that are specific to repo-updater as well as observability.
Access to the shared repo store is made by calling the `RepoStore()` or `ExternalServiceStore()` methods, which returns the shared store that reuses the existing db or transaction handle.
This PR doesn't plug the new store to repo-updater.
Fixes #14714

### CODE REVIEW

File: cmd/frontend/types/types.go [hunk 1]
```diff
@@ -192,6 +192,10 @@ func (r *Repo) Clone() *Repo {
 		return nil
 	}
 	clone := *r
+	if r.RepoFields != nil {
+		repoFields := *r.RepoFields
```
Comment (tsenart) [inline cmd/frontend/types/types.go:196]: This doesn't copy some of the fields which are reference types: https://github.com/sourcegraph/sourcegraph/blob/7282b18bbaad38c7de77ee9b265dd7a88931804f/cmd/frontend/types/types.go#L57-L61

File: cmd/repo-updater/internal/dbhelpers.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
+package internal
```
Comment (tsenart) [inline cmd/repo-updater/internal/dbhelpers.go:1]: I think this would fit in the dbutil package, no?

File: cmd/repo-updater/internal/store.go [hunk 3]
```diff
@@ -0,0 +1,834 @@
...
+)
+
+// A Store exposes methods to read and write repos and external services.
+type Store struct {
+	*basestore.Store
+
+	log     logging.ErrorLogger
+	metrics StoreMetrics
+	tracer  trace.Tracer
+	txtrace *trace.Trace
```
Comment (tsenart) [inline cmd/repo-updater/internal/store.go:30]: Can you add some comments here so it's obvious what each of these fields is used for?

File: cmd/repo-updater/internal/store.go [hunk 4]
```diff
@@ -0,0 +1,834 @@
...
+            external_service_repos sr
+            INNER JOIN external_services s ON s.id = sr.external_service_id
+        WHERE
+            s.namespace_user_id IS NOT NULL
+            AND s.deleted_at IS NULL
+            AND r.id = sr.repo_id
+            AND r.deleted_at IS NULL)
+`
+
+func (s *Store) EnqueueSyncJobs(ctx context.Context, ignoreSiteAdmin bool) (err error) {
```
Comment (tsenart) [inline cmd/repo-updater/internal/store.go:713]: Can we add some documentation here about this `ignoreSiteAdming` arg?

File: cmd/repo-updater/internal/store_test.go [hunk 5]
```diff
@@ -0,0 +1,1165 @@
...
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"github.com/inconshreveable/log15"
+	"github.com/opentracing/opentracing-go"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sourcegraph/sourcegraph/cmd/frontend/types"
+	"github.com/sourcegraph/sourcegraph/cmd/repo-updater/internal"
+	"github.com/sourcegraph/sourcegraph/internal/api"
+	"github.com/sourcegraph/sourcegraph/internal/conf"
+	"github.com/sourcegraph/sourcegraph/internal/db"
```
Comment (tsenart) [inline cmd/repo-updater/internal/store_test.go:24]: ST1019: package "github.com/sourcegraph/sourcegraph/internal/db" is being imported more than once (stylecheck)

File: cmd/repo-updater/internal/store_test.go [hunk 6]
```diff
@@ -0,0 +1,1165 @@
...
+		services := internaldb.MakeExternalServices()
+
+		if err := store.ExternalServiceStore().Upsert(ctx, services...); err != nil {
+			t.Fatalf("Upsert error: %v", err)
+		}
+
+		servicesPerKind := internaldb.ExternalServicesToMap(services)
+
+		repositories := types.Repos{
+			internaldb.MakeGithubRepo(servicesPerKind[extsvc.KindGitHub]),
```
Comment (tsenart) [inline cmd/repo-updater/internal/store_test.go:184]: Nice! These test fakes were out of control!

Comment (tsenart) [reply cmd/repo-updater/internal/store_test.go:184]: Should these be in the `types` package though? I don't think they have a lot to do with db.

File: cmd/repo-updater/internal/sync.go [hunk 7]
```diff
@@ -0,0 +1,22 @@
+package internal
+
+import "database/sql"
```
Comment (tsenart) [inline cmd/repo-updater/internal/sync.go:3]: I'd rename this file to sync_job.go

**Review by tsenart — APPROVED** [review]
Nice!

File: cmd/frontend/types/types.go (hunk 1 above)
Comment (asdine) [reply cmd/frontend/types/types.go:196]: Good catch!

File: cmd/repo-updater/internal/dbhelpers.go (hunk 2 above)
Comment (asdine) [reply cmd/repo-updater/internal/dbhelpers.go:1]: Yes good point

File: cmd/frontend/types/types.go (hunk 1 above)
Comment (asdine) [reply cmd/frontend/types/types.go:196]: Sources are copied but metadata isn't. However in the Update method above there is a comment saying that Clone is not expected to clone the medatata.

https://github.com/sourcegraph/sourcegraph/blob/6a14d3cd1595f6576e64a74c98834bdc80710886/cmd/frontend/types/types.go#L175-L178

I'd be in favor of keeping this behavior for now, what do you think?

Comment (tsenart) [reply cmd/frontend/types/types.go:196]: Yeah, SGTM.

File: cmd/repo-updater/internal/store.go [hunk 8]
```diff
@@ -0,0 +1,879 @@
...
+	return &Store{Store: s.Store.With(other)}
+}
+
+// RepoStore returns a db.ReposStore using the same database handle.
+func (s *Store) RepoStore() *internaldb.RepoStore {
+	return internaldb.Repos.With(s.Store)
+}
+
+// ExternalServiceStore returns a db.ExternalServiceStore using the same database handle.
+func (s *Store) ExternalServiceStore() *internaldb.ExternalServiceStore {
```
Comment (unknwon) [inline cmd/repo-updater/internal/store.go:130]: I've been waiting for it!!

**Pull request merged**